client/pool.go

```diff
@@ -0,0 +1,172 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"golang.org/x/net/proxy"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+)
+
+// ClientPool is an implementation of the APIClient interface which will handle
+// server failure, rotate servers, and retry API requests.
+type ClientPool struct {
+	*InsightClient
+	urls         []string
+	activeServer int
+	proxyDialer  proxy.Dialer
+	blockChan    chan Block
+	txChan       chan Transaction
+	httpClient   http.Client
+	cancelFunc   context.CancelFunc
+	client       *http.Client
+}
+
+// NewClientPool instantiates a new ClientPool object with the given server APIs
+func NewClientPool(urls []string, proxyDialer proxy.Dialer) (*ClientPool, error) {
+	for _, apiUrl := range urls {
+		u, err := url.Parse(apiUrl)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := validateScheme(u); err != nil {
+			return nil, err
+		}
+	}
+	return &ClientPool{
+		urls:        urls,
+		blockChan:   make(chan Block),
+		txChan:      make(chan Transaction),
+		proxyDialer: proxyDialer,
+	}, nil
+}
+
+// Start will attempt to connect to the first server URl. If it fails to
+// connect it will rotate through the servers to try to find one that works.
+func (p *ClientPool) Start() error {
+	for _, url := range p.urls {
+		client, err := NewInsightClient(url, p.proxyDialer)
+		if err != nil {
+			Log.Error(err)
+			continue
+		}
+		p.InsightClient = client
+		p.InsightClient.requestFunc = p.doRequest
+		ctx, cancel := context.WithCancel(context.Background())
+		p.cancelFunc = cancel
+		if p.client != nil {
+			p.httpClient = *p.client
+		}
+		go p.listenChans(ctx)
+		go p.connectWebsockets()
```

Comment:
Async call to connectWebsockets looks like a race condition in a failure case. Assume we return the ClientPool instance as having successfully started, but the websocket errors and begins rotating the server. The ClientPool can immediately execute a request which would panic intermittently depending on whether the rotation was completed or not.